src/metrics/mod.rs

```diff
@@ -180,18 +181,112 @@ impl RecentlyAccessedReleases {
     }
 }
 
-impl Metrics {
+impl InstanceMetrics {
+    pub(crate) fn gather(&self, pool: &Pool) -> Result<Vec<MetricFamily>, Error> {
+        self.idle_db_connections.set(pool.idle_connections() as i64);
+        self.used_db_connections.set(pool.used_connections() as i64);
+        self.max_db_connections.set(pool.max_size() as i64);
+
+        self.recently_accessed_releases.gather(self);
+        self.gather_system_performance();
+        Ok(self.registry.gather())
+    }
+
+    #[cfg(not(target_os = "linux"))]
+    fn gather_system_performance(&self) {}
+
+    #[cfg(target_os = "linux")]
+    fn gather_system_performance(&self) {
+        use procfs::process::Process;
+
+        let process = Process::myself().unwrap();
+        self.open_file_descriptors
+            .set(process.fd_count().unwrap() as i64);
+        self.running_threads
+            .set(process.stat().unwrap().num_threads);
+    }
+}
+
+fn metric_from_opts<T: MetricFromOpts + Clone + prometheus::core::Collector + 'static>(
+    registry: &prometheus::Registry,
+    metric: &str,
+    help: &str,
+    variable_label: Option<&str>,
+) -> Result<T, prometheus::Error> {
+    let mut opts = prometheus::Opts::new(metric, help).namespace("docsrs");
+
+    if let Some(label) = variable_label {
+        opts = opts.variable_label(label);
+    }
+
+    let metric = T::from_opts(opts)?;
+    registry.register(Box::new(metric.clone()))?;
+    Ok(metric)
+}
+
+#[derive(Debug)]
+pub struct ServiceMetrics {
+    pub queued_crates_count: IntGauge,
+    pub prioritized_crates_count: IntGauge,
+    pub failed_crates_count: IntGauge,
+    pub queue_is_locked: IntGauge,
+    pub queued_crates_count_by_priority: IntGaugeVec,
+    pub queued_cdn_invalidations_by_distribution: IntGaugeVec,
+
+    registry: prometheus::Registry,
+}
+
+impl ServiceMetrics {
+    pub fn new() -> Result<Self, prometheus::Error> {
```

Comment:
Is this just doing what the macro generates, or is there a reason the `InstanceMetrics` can use it but this one can't?

Reply:
I honestly don't fully remember the details. 

I think it was a combination: 
- the service metrics would have needed more special-casing in the macro (like cdn invalidation times), or would have needed a refactor
- the special-casing makes it harder to reason about the code IMO.